pkg/storegateway/series_chunks.go

```diff
@@ -365,24 +387,69 @@ func (c *loadingSeriesChunksSetIterator) Next() (retHasNext bool) {
 
 			err := c.chunkReaders.addLoad(chunk.blockID, chunk.ref, i, j)
 			if err != nil {
-				c.err = errors.Wrap(err, "preloading chunks")
-				return false
+				return nil, false, errors.Wrap(err, "preloading chunks")
 			}
 		}
 	}
 
 	// Create a batched memory pool that can be released all at once.
 	chunksPool := pool.NewSafeSlabPool[byte](chunkBytesSlicePool, chunkBytesSlabSize)
 
-	err := c.chunkReaders.load(nextSet.series, chunksPool, c.stats)
+	err = c.chunkReaders.load(nextSet.series, chunksPool, c.stats)
 	if err != nil {
-		c.err = errors.Wrap(err, "loading chunks")
-		return false
+		return nil, false, errors.Wrap(err, "loading chunks")
 	}
 
 	nextSet.chunksReleaser = chunksPool
-	c.current = nextSet
-	return true
+
+	if c.ignoreNativeHistograms {
+		nextSet.series = filterNativeHistogramChunks(nextSet.series)
+		if len(nextSet.series) == 0 {
+			nextSet.release()
+			return nil, true, nil
+		}
+	}
+
+	return &nextSet, false, nil
+}
+
+func filterNativeHistogramChunks(seriesEntries []seriesEntry) []seriesEntry {
+	seriesWriteIdx := 0
+	for i, series := range seriesEntries {
+		chksWriteIdx := 0
+		for j, chk := range series.chks {
+			// The bucketChunkReader will skip populating native histogram chunks when ignoreNativeHistograms is true,
+			// leaving chk.Raw as nil.
+			if chk.Raw == nil {
+				chk.Reset()
+				continue
+			}
+
+			series.chks[chksWriteIdx] = series.chks[j]
+			if series.refs != nil {
+				// Not all callers will have series.refs set, but when it is set we expect refs and chks to have a
+				// one-to-one relationship.
+				series.refs[chksWriteIdx] = series.refs[j]
+			}
+
+			chksWriteIdx++
+		}
+		// Any remaining chunks are to be discarded but we need to reset them first.
+		for k := chksWriteIdx; k < len(series.chks); k++ {
+			series.chks[k].Reset()
```

Comment:
Why calling Reset() here? I don't think it's necessary. Could you explain why is it required?

Reply:
Yep! I was trying to figure out why the "loads single set from a single block with all ignored histogram chunks" [test](https://github.com/grafana/mimir/blob/650beb3d338b3b1f13e4269fff5abfefc62510c6/pkg/storegateway/series_chunks_test.go#L639) would fail every time it was run directly after the "loads single set from single block with some ignored histogram chunks" [test](https://github.com/grafana/mimir/blob/650beb3d338b3b1f13e4269fff5abfefc62510c6/pkg/storegateway/series_chunks_test.go#L605) and it ended up being that XOR chunks from the "some ignored histogram chunks" test would end up being processed in the "all ignored histogram chunks" test because the tests all use the same `seriesChunksSlicePool`.

I then realized that `seriesChunksSet.release()` resets all chunks [here](https://github.com/grafana/mimir/blob/650beb3d338b3b1f13e4269fff5abfefc62510c6/pkg/storegateway/series_chunks.go#L123) but the chunks we discard in `filterNativeHistogramChunks` would be unreachable at that point so they'd never get reset.

I suspect this really only matters for this bit of ignore native histogram code because normally, a chunk would get populated over the old chunk anyways but because we [skip doing anything for native histogram chunks at all when that flag is active](https://github.com/grafana/mimir/blob/650beb3d338b3b1f13e4269fff5abfefc62510c6/pkg/storegateway/bucket_chunk_reader.go#L169-L171) an old chunk can then show up in the slice of chunks.

Reply:
in this case is it possible to change `seriesChunksSet.release()` to reset all chunks in its capacity instead of only the ones in its length?

Reply:
Good idea--yea, I don't see why not!

Reply:
Oop, actually I think we cannot--that would work for the chunks that are left on the end, but some chunks are also just written over so they wouldn't end up reachable when `seriesChunksSet.release()` is called.

Reply:
Ohh I see it now--thanks for the explanation! Ok yea, making that change to reseting all chunks in capacity instead of just the length in `seriesChunksSet.release()`.